### DISCUSSION TITLE
AsyncNotifier. Should build method  also return AsyncLoading?

### DISCUSSION THREAD

**Discussion by Fraa-124** [rrousselGit/riverpod#1820] [Q&A] [answered]
Lets say,  my state depends on AuthProvider, I'm watching this provider for changes in build method. When the user signs out, its value is null, and I want to return AsyncLoading in the build method.

For the sake of simplicity I omit relevant parts
```
class ContactState {
Auth auth,
...
}

class ContactDetailsAsyncNotifier extends AsyncNotifier<ContactState> {
  @override
  FutureOr<ContactState> build() {
    final auth = ref.watch(authProvider);

    //if (auth == null) {
    //  return AsyncLoading();
    //}

    return ContactState(auth);
  }
}
```

How could I achieve something like this?

**Comment by rrousselGit** [comment, marked as answer]
You can return `this.future`

```dart

class ContactDetailsAsyncNotifier extends AsyncNotifier<ContactState> {
  @override
  FutureOr<ContactState> build() {
    final auth = ref.watch(authProvider);

    if (auth == null) {
      return future;
    }

    return ContactState(auth);
  }
}
```

**Reply by Fraa-124** [reply]
It works, this is black magic for me 🤯. Thank you Remi.

**Reply by rrousselGit** [reply]
The idea is that `future` resolves when a data/error is emitted. So if there's no data/error, it'll wait for it.

This is typically used inside change methods like "setUserName", to obtain the data before changing it. But it can be used this way too
